### ISSUE TITLE
False negative ``undefined-variable`` after a caught exception shadowing a variable name

### ISSUE THREAD

**Issue by pylint-bot** [PyCQA/pylint#157]
Originally reported by: **BitBucket: [moxian](http://bitbucket.org/moxian), GitHub: @moxian?**

---

Pylint'ing snippet

```python
def main():
    exc = None
    try:
        [][1] = 12
    except IndexError as exc:
        pass
    if exc is not None:
        raise exc


main()

```

says

```
Raising NoneType while only classes, instances or string are allowed (raising-bad-type)
```

when in fact None will never be attempted to be raised

Python 2

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/157

**Comment by Pierre-Sassoulas** [comment]
Given:
```python
# pylint: disable=missing-docstring


def main():
    exc = None
    print(exc)
    try:
        [][1] = 12
    except IndexError as exc:
        print(exc)
    if exc is not None:
        raise exc


main()

```
Pylint does not raise any errors but running the code gives:
```
None
list assignment index out of range
Traceback (most recent call last):
  File "a.py", line 15, in <module>
    main()
  File "a.py", line 11, in main
    if exc is not None:
UnboundLocalError: local variable 'exc' referenced before assignment
```
So it actually became a false negative for undefined variable in python 3.
